### DISCUSSION TITLE
zh/tasks/detect/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9845] [Docs]
# zh/tasks/detect/

YOLOv8 的官方文档，由Ultralytics 提供。了解如何以各种格式训练、验证、预测和导出模型。包括详细的性能统计。

https://docs.ultralytics.com/zh/tasks/detect/

**Comment by LGxaz** [comment]
Traceback (most recent call last):
  File "D:\Anaconda\anaconda3\envs\yolo_py38\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\Anaconda\anaconda3\envs\yolo_py38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\Anaconda\anaconda3\envs\yolo_py38\Scripts\yolo.exe\__main__.py", line 7, in <module>
  File "F:\yolov8\ultralytics\ultralytics\cfg\__init__.py", line 582, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "F:\yolov8\ultralytics\ultralytics\engine\model.py", line 668, in train
    self.trainer.train()
  File "F:\yolov8\ultralytics\ultralytics\engine\trainer.py", line 198, in train
    self._do_train(world_size)
  File "F:\yolov8\ultralytics\ultralytics\engine\trainer.py", line 418, in _do_train
    self.metrics, self.fitness = self.validate()
  File "F:\yolov8\ultralytics\ultralytics\engine\trainer.py", line 557, in validate
    metrics = self.validator(self)
  File "D:\Anaconda\anaconda3\envs\yolo_py38\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\yolov8\ultralytics\ultralytics\engine\validator.py", line 195, in __call__
    stats = self.get_stats()
  File "F:\yolov8\ultralytics\ultralytics\models\yolo\detect\val.py", line 170, in get_stats
    stats = {k: torch.cat(v, 0).cpu().numpy() for k, v in self.stats.items()}  # to numpy
  File "F:\yolov8\ultralytics\ultralytics\models\yolo\detect\val.py", line 170, in <dictcomp>
    stats = {k: torch.cat(v, 0).cpu().numpy() for k, v in self.stats.items()}  # to numpy
RuntimeError: torch.cat(): expected a non-empty list of Tensors

**Reply by pderrenger** [reply]
Hey there 👋, it looks like you're encountering a `RuntimeError` related to `torch.cat()`: it expects a non-empty list of tensors but received an empty one. This error often occurs when the model doesn't detect any objects in the validation set, leading to an empty list of results being passed to `torch.cat()`.

To troubleshoot, consider the following steps:
- Ensure your dataset is properly loaded and accessible to YOLOv8 during validation.
- Check if the provided `data` path in your config file points to valid dataset directories.
- Temporarily reduce the `conf` threshold in your training script to see if it starts detecting objects. For example, lowering it to `0.01` might help:
  
  ```bash
  yolo train data=coco128.yaml model=yolov8n.yaml conf=0.01
  ```

If the issue persists, it could be helpful to share more details about your setup, including the dataset and configuration files, so we can offer more targeted advice! 😊

**Reply by LGxaz** [reply]
感谢回复，已经解决，问题在于数据集xml转yolo失败。

**Reply by pderrenger** [reply]
太棒了，很高兴听到您解决了问题！🎉 数据集转换问题确实会导致一些头疼的错误。如果未来您有任何关于模型训练或部署的疑问，或者需要关于数据准备的更多建议，随时欢迎您回来提问。🙂

下次遇到XML转YOLO格式失败时，可以尝试使用我们提及的[JSON2YOLO工具](https://github.com/ultralytics/json2yolo)来简化流程。这个工具能帮助您轻松转换不同格式的数据集。

祝您在接下来的项目中一切顺利！

**Comment by onetree10** [comment]
Does the YOLOv8 pre trained model for object detection support training its own dataset (the dataset format is converted from polygon annotated JSON format to YOLO format)

**Reply by glenn-jocher** [reply]
Yes, the YOLOv8 pre-trained model supports training on your own dataset, even if it was originally in a different annotation format like polygon annotated JSON. You'll need to convert your dataset to the YOLO format first. Once your dataset is in YOLO format, you can use it for training by specifying your dataset in the `data` YAML file and then proceeding with training as usual.

Here's a quick example on how to train with your custom dataset after conversion:

```bash
# Load your model and train on your custom dataset
yolo detect train data=your_dataset.yaml model=yolov8n.pt epochs=100 imgsz=640
```

Make sure your `your_dataset.yaml` correctly points to your training and validation data in the YOLO format. Happy training! 🚀

**Reply by onetree10** [reply]
I have now annotated a batch of data (in JSON format) using the polygon annotation tool, which I want to use for object detection tasks. Does YOLOv8's object detection pre-trained model support polygon-annotated data? YOLOv5 does not support polygon-annotated data. (Or YOLOv8 supports polygon annotation data, pay attention to the separation from the instance when converting json to YOLO?)

**Reply by glenn-jocher** [reply]
Hey there! 👋 YOLOv8, similar to YOLOv5, fundamentally works with bounding boxes for object detection and doesn't natively support polygon annotations directly in its training process. But, don't worry! You can convert your polygon-annotated data into bounding boxes following a simple approach:

1. Convert your polygon annotations to the bounding box format by calculating the minimum and maximum coordinates enclosing your polygon.

For example, if you have a polygon annotation in JSON:
```json
{
  "shapes": [
    {
      "label": "YourObject",
      "points": [[x1, y1], [x2, y2], [x3, y3], [x4, y4]]
    }
  ]
}
```
Convert this to a bounding box by finding the min and max of `x` and `y` coordinates.

2. Once converted to bounding boxes, format your data in the YOLO format:
```
<object-class> <x_center> <y_center> <width> <height>
```

Make sure to normalize `x_center`, `y_center`, `width`, and `height` by the image width and height to convert them into relative coordinates.

Remember, the transition from polygons to bounding boxes might not capture the exact essence of very irregular shapes but works well for a wide range of applications. Train your model as usual after conversion! 🚀

Let us know if you need any further assistance. Happy training!

**Reply by onetree10** [reply]
JSON2YOLO the "general_json2yolo" and "labelbox_json2yolo" in the tool, the two scripts are respectively the JSON format of the dataset of the task and the YOLO format of the dataset of the task (for example, the format of the JSON dataset of instance segmentation is accurately changed to the YOLO dataset format of the target detection).

**Reply by glenn-jocher** [reply]
Hey there! 👋 It looks like you're exploring the tools `general_json2yolo` and `labelbox_json2yolo` for converting JSON dataset formats to YOLO format, specifically for instance segmentation to object detection tasks. That's an exciting step towards customizing your models! 🚀

When converting, ensuring the bounding boxes derived from your polygon annotations accurately encompass your objects of interest is crucial. This conversion process retains essential object details while complying with YOLO's detection format.

Here's a quick Python code snippet illustrating how you might convert JSON polygon annotations to YOLO format bounding boxes:

```python
# Assuming 'annotations' is your list of polygon annotations from JSON
for annotation in annotations:
    # Find the min and max coordinates to form a bounding box
    x_coordinates = [point[0] for point in annotation['points']]
    y_coordinates = [point[1] for point in annotation['points']]
    bbox = [min(x_coordinates), min(y_coordinates), max(x_coordinates), max(y_coordinates)]
    
    # Convert to YOLO format (x_center, y_center, width, height)
    bbox_yolo = convert_to_yolo_format(bbox, image_width, image_height)
    
    # Save or use the bbox_yolo for your YOLO training data
```

Remember, converting annotations formats might require some tweaks depending on the specifics of your dataset. Always check to ensure the converted bounding boxes are accurate and represent your objects well. 

Feel free to reach out if you have further questions or need more help. Happy training! 🤗

**Comment by gulanyun** [comment]
The results of the training are not available, metrics/mAP50(B),    metrics/mAP50-95(B),always too small

**Reply by glenn-jocher** [reply]
@gulanyun it seems like your training results are showing low mAP scores. To improve this, consider checking your dataset quality, ensuring proper annotation, and adjusting hyperparameters like learning rate or batch size. For more detailed guidance, you might want to explore the Ultralytics documentation on training strategies.

**Comment by xiao5676wen** [comment]
对视频进行检测时，设定参数save=Ture，但是保存的时一个0字节的avi格式视频，我怎样才能顺利的保存视频，而且我能不能用参数指定保存的视频格式？

**Reply by xiao5676wen** [reply]
检测视频后，保存检测视频却保存出一个0字节的avi格式视频！我想顺利保存出正确的视频，并且希望能用预测时的参数指定我想保存的视频格式。

**Reply by glenn-jocher** [reply]
@xiao5676wen 在视频检测保存为0字节的情况下，可能与编码器设置或输出路径权限有关。请确保 `ffmpeg` 安装正确，因为它是保存视频的依赖工具。同时，参数 `save` 无法直接指定格式，默认保存为 `.mp4`。如需其他格式，建议后处理时用工具如 `ffmpeg` 转换格式。运行时，检查日志是否有错误提示，帮助排查问题。

**Comment by Leeeeah** [comment]
如果我要修改训练时iou的值，我该如何修改呢？我查了好久都查不到正确的修改路径？

**Reply by glenn-jocher** [reply]
@Leeeeah 要修改训练中的 IoU 值，您可以通过调整配置参数 `iou_t`，它表示正样本的 IoU 阈值。

例如，在 Python 代码中，您可以这样设置：
```python
model.train(data="dataset.yaml", epochs=50, iou_t=0.7)
```

或者使用命令行界面（CLI）：
```
yolo detect train data=dataset.yaml epochs=50 iou_t=0.7
```

更多参数详情可参考[Ultralytics配置文档](https://docs.ultralytics.com). 如果您有其他问题，欢迎继续咨询！

**Comment by legend-zhen** [comment]
Can the weight file be converted from .pt format to .pth format? How should it be converted?please teach me.Thank you！

**Reply by glenn-jocher** [reply]
@legend-zhen to convert a .pt Ultralytics YOLO model to .pth format, simply rename the file (`mv model.pt model.pth`) as both formats store PyTorch state dicts. For official export options like ONNX or TensorRT, see the Export documentation. 

For commercial use, remember AGPL-3.0 requires open-sourcing derivatives or obtaining an Enterprise License from Ultralytics.

**Comment by legend-zhen** [comment]
Hello, I would like to ask how to use YOLOv8 as the object detector in a top-down paradigm for pose estimation tasks. For example, how can I combine the YOLOv8 network with the HRNet keypoint detection network? How should this be implemented? Thank you.

**Reply by glenn-jocher** [reply]
@legend-zhen to implement YOLOv8 with HRNet in a top-down pose estimation pipeline: 

1. Use YOLOv8 Detect to get bounding boxes for persons
2. Crop detected regions and pass each to HRNet for keypoint detection
3. Process results sequentially or batch-process crops

Here's a minimal implementation sketch:
```python
from ultralytics import YOLO
import hrnet_lib  # Your HRNet setup

detector = YOLO('yolov8n.pt')
pose_model = hrnet_lib.load_model()

results = detector(source_image)
for box in results[0].boxes.xyxy:  # Get detection boxes
    crop = source_image.crop(box)
    keypoints = pose_model(crop)  # HRNet inference
```

For complete guidance, see our Pose Estimation documentation and Python API examples.

**Comment by legend-zhen** [comment]
Hello, I am using YOLO11 for multi-class object detection (around 40 categories). Currently, the detection boxes perform well, but the classification accuracy is relatively poor. I would like to ask, what aspects of the network should be improved in order to enhance its classification performance? Thank you.

**Reply by glenn-jocher** [reply]
@legend-zhen Hello! Improving classification accuracy while detection boxes are good often involves focusing on aspects that enhance feature representation for distinguishing between classes. Here are a few areas you could explore:

1.  **Data Augmentation:** Review and potentially increase augmentations that affect appearance more than geometry, like color space adjustments (`hsv_h`, `hsv_s`, `hsv_v`).
2.  **Hyperparameter Tuning:** Consider slightly increasing the classification loss gain hyperparameter (`cls`) to give classification more weight during training. You might also experiment with training for more epochs.
3.  **Model Size:** If you are using a smaller model (like YOLO11n or YOLO11s), trying a larger one (YOLO11m, YOLO11l, YOLO11x) could provide the capacity needed to better differentiate between your 40 classes.
4.  **Dataset Review:** Ensure your dataset has sufficient and balanced examples for each class, especially for classes that are commonly confused. High-quality, diverse images are key.

You can find more details on adjusting hyperparameters and training strategies in our [Tips for Model Training guide](https://docs.ultralytics.com/guides/model-training-tips/). Good luck with your improvements!

**Comment by likesnow97** [comment]
YOLO是否实现了将预测结果导出为x-anylabeling可用的json格式接口？

**Comment by legend-zhen** [comment]
Hello, I would like to use the weight file obtained after modifying the YOLO11 model to provide a checkbox for the top-down pose estimation algorithm. Why can it only run normally under the official weight file(yolo11.pt)? When using the weight of the YOLO11 structure improved by myself, the following error message appears. Please ask what should be done to use the weight file trained by myself.
D:\Anaconda\envs\openmmlab\python.exe D:\openmmlab\mmpose-11\topdown_demo_with_YOLO1111.py 
D:\Anaconda\envs\openmmlab\lib\site-packages\mmengine\utils\package_utils.py:48: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import DistributionNotFound, get_distribution
WARNING ⚠️ D:\YOLO\YOLO11\ultralytics-main\runs\detect\train8\weights\best.pt appears to require 'ultralytics.nn.add_modules', which is not in Ultralytics requirements.
AutoInstall will run now for 'ultralytics.nn.add_modules' but this feature will be removed in the future.
Recommend fixes are to train a new model using the latest 'ultralytics' package or to run a command with an official Ultralytics model, i.e. 'yolo predict model=yolov8n.pt'
Traceback (most recent call last):
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\nn\tasks.py", line 837, in torch_safe_load
    ckpt = torch.load(file, map_location="cpu")
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\utils\patches.py", line 86, in torch_load
    return _torch_load(*args, **kwargs)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 712, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 1049, in _load
    result = unpickler.load()
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 1042, in find_class
    return super().find_class(mod_name, name)
ModuleNotFoundError: No module named 'ultralytics.nn.add_modules'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\openmmlab\mmpose-11\topdown_demo_with_YOLO1111.py", line 316, in <module>
    main()
  File "D:\openmmlab\mmpose-11\topdown_demo_with_YOLO1111.py", line 196, in main
    model_yolo11 = YOLO(args.det_checkpoint)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\models\yolo\model.py", line 23, in __init__
    super().__init__(model=model, task=task, verbose=verbose)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\engine\model.py", line 145, in __init__
    self._load(model, task=task)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\engine\model.py", line 285, in _load
    self.model, self.ckpt = attempt_load_one_weight(weights)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\nn\tasks.py", line 910, in attempt_load_one_weight
    ckpt, weight = torch_safe_load(weight)  # load ckpt
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\nn\tasks.py", line 857, in torch_safe_load
    ckpt = torch.load(file, map_location="cpu")
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\ultralytics\utils\patches.py", line 86, in torch_load
    return _torch_load(*args, **kwargs)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 712, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 1049, in _load
    result = unpickler.load()
  File "D:\Anaconda\envs\openmmlab\lib\site-packages\torch\serialization.py", line 1042, in find_class
    return super().find_class(mod_name, name)
ModuleNotFoundError: No module named 'ultralytics.nn.add_modules'

**Reply by glenn-jocher** [reply]
@legend-zhen Thanks for the traceback. Your checkpoint was saved with custom code (`ultralytics.nn.add_modules`) from a modified Ultralytics repo, but you’re loading it with the pip-installed package where that module doesn’t exist. Fix by either installing your modified repo into the same environment, or exporting to a format that doesn’t require the original Python classes:

```bash
# Option A: load with the same modified code
pip uninstall -y ultralytics
pip install -e path\to\your_modified_ultralytics

# Option B: export once from the environment where it loads, then use the artifact
yolo export model="D:\YOLO\YOLO11\ultralytics-main\runs\detect\train8\weights\best.pt" format=onnx
```

Please also run and paste the text output of:
```bash
yolo checks
```
to confirm you’re not mixing a local repo and a pip install; compare the Ultralytics install path it prints with the paths in your traceback. Export formats and usage are summarized in the [Export guide](https://docs.ultralytics.com/usage/cli/) and general troubleshooting tips are in [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/).

**Comment by MrBug1024** [comment]
I have been using YOLO for many years, but I have always been looking forward to the official release of a face recognition solution. Because the current open-source solutions in the face recognition field, whether in terms of performance or subsequent enhanced training, are not as simple as training a custom object detection model with YOLO. This is what shocks me. Therefore, I still hope that the official team can provide a convenient solution like object detection in the fields of face recognition and OCR!

**Comment by VeryOnline** [comment]
你好，yolo的开发者，我使用yolo的CLI转换ONNX模型后，得到的onnx模型输出是一个float32[1,300,6]的输出，这里的输出是如何定义的？我没有找到相关的定义，能帮我解释一下这个输出么？

**Reply by glenn-jocher** [reply]
你好，你看到的 `float32[1,300,6]` 一般是导出时启用了 `nms=True`（或导出为端到端后处理输出）得到的结果：`300` 表示最多保留的检测数（`max_det`），每个检测的 6 个值通常对应 `x1, y1, x2, y2, conf, cls`（`cls` 会以 float 形式存储）。如果你想拿到未做 NMS 的“原始 head 输出”用于自定义后处理，需要用 `nms=False` 重新导出，并按检测模型常见的 `(B, 4+num_classes, num_predictions)` 结构解析，参考 [导出文档里的输出张量说明](https://docs.ultralytics.com/modes/export/)（也可对照 [目标检测任务文档](https://docs.ultralytics.com/tasks/detect/) 的用法确认你导出的模型类型）。

```bash
yolo export model=yolo26n.pt format=onnx nms=True
yolo export model=yolo26n.pt format=onnx nms=False
```

如果方便的话，把你实际的导出命令（包含 `imgsz` / `nms` 等参数）和 ONNX 的 input/output 名称贴在这个讨论里，我们可以更精确地确认每一维的含义和坐标尺度。
